pkg/meta/tkv.go

```diff
@@ -2514,74 +2514,142 @@ func (m *kvMeta) DumpMeta(w io.Writer, root Ino) (err error) {
+			if err == nil {
+				inode := e.Attr.Inode
+				e.Parents = append(parents[inode], parent)
+				parents[inode] = e.Parents
+				if len(e.Parents) == 1 {
+					if inode > 1 && inode != TrashInode {
+						cs.UsedSpace += align4K(e.Attr.Length)
+						cs.UsedInodes += 1
+					}
+					if inode < TrashInode {
+						if cs.NextInode <= int64(inode) {
+							cs.NextInode = int64(inode) + 1
+						}
+					} else {
+						if cs.NextTrash < int64(inode)-TrashInode {
+							cs.NextTrash = int64(inode) - TrashInode
+						}
+					}
 				}
-				slices := make([]byte, 0, sliceBytes*len(c.Slices))
-				m.Lock()
-				for _, s := range c.Slices {
-					slices = append(slices, marshalSlice(s.Pos, s.Chunkid, s.Size, s.Off, s.Len)...)
-					refs[string(m.sliceKey(s.Chunkid, s.Size))]++
-					if cs.NextChunk <= int64(s.Chunkid) {
-						cs.NextChunk = int64(s.Chunkid) + 1
+			}
+		case "chunks":
+			err = dec.Decode(&e.Chunks)
+			if err == nil && len(e.Parents) == 1 {
+				for _, c := range e.Chunks {
+					for _, s := range c.Slices {
+						refs[chunkKey{s.Chunkid, s.Size}]++
+						if cs.NextChunk <= int64(s.Chunkid) {
+							cs.NextChunk = int64(s.Chunkid) + 1
+						}
 					}
 				}
-				m.Unlock()
-				tx.set(m.chunkKey(inode, c.Index), slices)
 			}
-		} else if attr.Typ == TypeDirectory {
-			attr.Length = 4 << 10
-			for _, c := range e.Entries {
-				tx.set(m.entryKey(inode, string(unescape(c.Name))), m.packEntry(typeFromString(c.Attr.Type), c.Attr.Inode))
+		case "entries":
+			e.Entries = make(map[string]*DumpedEntry)
+			_, err = dec.Token()
+			if err == nil {
+				for dec.More() {
+					var n json.Token
+					n, err = dec.Token()
+					if err != nil {
+						break
+					}
+					var child *DumpedEntry
+					child, err = m.decodeEntry(dec, e.Attr.Inode, cs, parents, refs, kv, show_progress)
+					if err != nil {
+						break
+					}
+					e.Entries[n.(string)] = &DumpedEntry{
+						Attr: &DumpedAttr{
+							Inode: child.Attr.Inode,
+							Type:  child.Attr.Type,
+						},
+					}
+				}
+				if err == nil {
+					var t json.Token
+					t, err = dec.Token()
+					if err == nil && t != json.Delim('}') {
+						err = fmt.Errorf("unexpected %v", t)
+					}
+				}
 			}
-		} else if attr.Typ == TypeSymlink {
-			symL := unescape(e.Symlink)
-			attr.Length = uint64(len(symL))
-			tx.set(m.symKey(inode), []byte(symL))
-		}
-		m.Lock()
-		if inode > 1 && inode != TrashInode {
-			cs.UsedSpace += align4K(attr.Length)
-			cs.UsedInodes += 1
+		case "symlink":
+			err = dec.Decode(&e.Symlink)
+		case "xattrs":
+			err = dec.Decode(&e.Xattrs)
 		}
-		if inode < TrashInode {
-			if cs.NextInode <= int64(inode) {
-				cs.NextInode = int64(inode) + 1
-			}
-		} else {
-			if cs.NextTrash < int64(inode)-TrashInode {
-				cs.NextTrash = int64(inode) - TrashInode
-			}
+		if err != nil {
+			return nil, fmt.Errorf("decode %v: %s", name, err)
 		}
-		m.Unlock()
+	}
+	if len(e.Parents) == 1 {
+		_ = m.loadEntry(&e, cs, parents, refs, kv)
```

Comment:
In this case, we use the attribute of the first link, not the newest one any more